core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -0,0 +1,127 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.encryption;
+
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import javax.crypto.AEADBadTagException;
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+public class Ciphers {
+
+  public static class AesGcmEncryptor {
+    public static final int NONCE_LENGTH = 12;
+    public static final int GCM_TAG_LENGTH = 16;
+    public static final int GCM_TAG_LENGTH_BITS = 8 * GCM_TAG_LENGTH;
+
+    private final SecretKeySpec aesKey;
+    private final Cipher cipher;
+    private final SecureRandom randomGenerator;
+
+    public AesGcmEncryptor(byte[] keyBytes) {
+      int keyLength = keyBytes.length;
+      if (!(keyLength == 16 || keyLength == 24 || keyLength == 32)) {
+        throw new IllegalArgumentException("Wrong key length " + keyLength);
+      }
+      this.aesKey = new SecretKeySpec(keyBytes, "AES");
+
+      try {
+        this.cipher = Cipher.getInstance("AES/GCM/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to create GCM cipher", e);
+      }
+
+      this.randomGenerator = new SecureRandom();
+    }
+
+    public byte[] encrypt(byte[] plainText, byte[] aad) {
+      byte[] nonce = new byte[NONCE_LENGTH];
+      randomGenerator.nextBytes(nonce);
+      int cipherTextLength = NONCE_LENGTH + plainText.length + GCM_TAG_LENGTH;
+      byte[] cipherText = new byte[cipherTextLength];
+
+      try {
+        GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce);
+        cipher.init(Cipher.ENCRYPT_MODE, aesKey, spec);
+        if (null != aad) {
+          cipher.updateAAD(aad);
+        }
+        cipher.doFinal(plainText, 0, plainText.length, cipherText, NONCE_LENGTH);
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to encrypt", e);
+      }
+
+      // Add the nonce
+      System.arraycopy(nonce, 0, cipherText, 0, NONCE_LENGTH);
+
+      return cipherText;
+    }
+  }
+
+  public static class AesGcmDecryptor {
+    private final SecretKeySpec aesKey;
+    private final Cipher cipher;
+
+    public AesGcmDecryptor(byte[] keyBytes) {
+      int keyLength = keyBytes.length;
+      if (!(keyLength == 16 || keyLength == 24 || keyLength == 32)) {
+        throw new IllegalArgumentException("Wrong key length " + keyLength);
+      }
+
+      this.aesKey = new SecretKeySpec(keyBytes, "AES");
+
+      try {
+        this.cipher = Cipher.getInstance("AES/GCM/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to create GCM cipher", e);
+      }
+    }
+
+    public byte[] decrypt(byte[] ciphertext, byte[] aad)  {
+      int plainTextLength = ciphertext.length - AesGcmEncryptor.GCM_TAG_LENGTH - AesGcmEncryptor.NONCE_LENGTH;
```

Comment:
Nit: How about moving `AesGcmEncryptor.GCM_TAG_LENGTH`, `AesGcmEncryptor.NONCE_LENGTH`, `AesGcmEncryptor.GCM_TAG_LENGTH_BITS` under the class `Ciper`? So that class `AesGcmDecryptor` doesn't invoke the static methods in class `AesGcmEncryptor`.